### PR TITLE
Split out `AttributeAccessInterface.h` into separate headers/cpp for each underlying class

### PR DESCRIPTION
**Pull request by andy31415**
### Changes

- separate out `AttributeValueEncoder`, `AttributeValueDecoder` and `AttributeReportBuilder`
- Update includes for files that only generally want the encoder and not the entire attribute access interface

### CODE REVIEW

File: src/app/AttributeReportBuilder.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/*
+ *
+ *    Copyright (c) 2021-2022 Project CHIP Authors
```
Comment (bzbarsky-apple) [inline src/app/AttributeReportBuilder.h:3]: Why is the .cpp copyright 2024 but this is 2021-2022?  Seems like they should have more or less consistent copyright dates (probably whatever year this bit was added to 2024).

File: src/app/AttributeValueDecoder.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
```
Comment (bzbarsky-apple) [inline src/app/AttributeValueDecoder.h:2]: Should be `N-2024` where N is the year this class was added, I expect.

File: src/app/AttributeValueEncoder.cpp [hunk 3]
```diff
@@ -1,6 +1,5 @@
 /*
- *
- *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2024 Project CHIP Authors
```
Comment (bzbarsky-apple) [inline src/app/AttributeValueEncoder.cpp:2]: Should be `N-2024`?

File: src/app/AttributeValueEncoder.h [hunk 4]
```diff
@@ -0,0 +1,286 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
```
Comment (bzbarsky-apple) [inline src/app/AttributeValueEncoder.h:2]: See comments above.

**Review by bzbarsky-apple — APPROVED** [review]
Thank you for doing this!  Makes things much more readable, if nothing else.

**Review by yunhanw-google — APPROVED** [review]

**Comment by yunhanw-google** [comment]
Thanks for this split for better readability!

**Pull request merged**